### PR TITLE
 Community Features Implementation

### PR DESCRIPTION
**Pull request by mubarak23**

### CODE REVIEW

File: src/community/community.cairo [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+        }
+
+        fn add_community_mods(
+            ref self: ComponentState<TContractState>, community_id: u256, moderator: ContractAddress
+        ) {
+            // community owner should be able to set a mod 
+            let community_owner = get_caller_address();
+            let owner_community_id = self.communities_owner.read(community_owner);
+            let community = self.communities.read(community_id);
+            assert(owner_community_id != community_id, "Not Community Owner");
```
Comment (Darlington02) [inline src/community/community.cairo:215]: I'd prefer you use the community_id to get it's owner, and then compare against `get_caller_address`

File: src/community/community.cairo [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+            community_id: u256,
+            profile: ContractAddress,
+            ban_status: bool
+        ) {
+            let community_owner = get_caller_address();
+
+            // only community owner can take this action
+            let community = self.communities.read(community_id);
+
+            assert(commuity.community_owner == community_owner, "Not Community Owner");
```
Comment (Darlington02) [inline src/community/community.cairo:292]: not just community owners, but community mods should be able to ban users

File: src/community/community.cairo [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+            let comminuty_main_id = self.communities_owner.read(community_owner);
+            assert(comminuty_main_id == community_id, "Not Community Owner");
+            let community_details = self.communities.read(community_id);
+            let community_gate_keep_details = CommunityGateKeepDetails {
+                community_id: comminuty_main_id,
+                gate_keep_type: gatekeep_type,
+                community_nft_address: community_details.community_nft_address,
+                permissioned_addresses: permissioned_addresses,
+            };
+
```
Comment (Darlington02) [inline src/community/community.cairo:365]: TODO:

- If gatekeep_type is PermissionedGating, let's iterate through permissioned_addresses and store in `gatekeep_permissioned_addresses`.
- If it's PaidGating let's take in the entry_fee and add it to the `CommunityGateKeep` struct.

File: src/community/community.cairo [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+                    }
+                );
+        }
+        // IN_COMPLETE
+        fn gatekeep(
+            ref self: ComponentState<TContractState>,
+            community_id: u256,
+            gate_keep_type: GateKeepType,
+            nft_contract_address: ContractAddress,
+            permissioned_addresses: Vec<ContractAddress>
```
Comment (Darlington02) [inline src/community/community.cairo:352]: permissioned_addresses should rather be an array

File: src/community/community.cairo [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+        ) {
+            let community = self.communities.read(community_id);
+            if (!community) {
+                panic('Community does not exist');
+            }
+
+            // only owner can upgrade community
+            assert(commuity.community_owner == community_owner, "Not Community Owner");
+            community.community_type = upgrade_type;
+            self.communities.write(community_id, community);
```
Comment (Darlington02) [inline src/community/community.cairo:334]: let's also update premium_subscription_status to true

File: src/community/community.cairo [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+        ) -> bool {
+            let community_members = self.community_member.read(profile);
+            let memeber_details = community_members
+                .iter()
+                .find((community_member), community_member.profile_address == profile);
+            memeber_details.ban_status
+        }
+        fn is_premium_community(self: @TState, community_id: u256) -> (bool, CommunityType) {
+            let community = self.communities.read(community_id);
+            (community.community_type, community.community_premium_status)
```
Comment (Darlington02) [inline src/community/community.cairo:432]: return type seems exchanged?

**Review by Darlington02 — CHANGES_REQUESTED** [review]

File: src/community/community.cairo (hunk 3 above)
Comment (mubarak23) [reply src/community/community.cairo:365]: for this 

`If it's PaidGating let's take in the entry_fee and add it to the CommunityGateKeep struct.`

the function will need an extra argument, called `entry_fee` @Darlington02

**Pull request merged**